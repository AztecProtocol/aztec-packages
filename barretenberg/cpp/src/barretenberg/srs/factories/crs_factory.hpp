#pragma once
#include "barretenberg/common/mem.hpp"
#include "barretenberg/ecc/curves/bn254/bn254.hpp"
#include "barretenberg/ecc/curves/bn254/g1.hpp"
#include "barretenberg/ecc/curves/bn254/g2.hpp"
#include "barretenberg/ecc/curves/grumpkin/grumpkin.hpp"
#include <cstddef>

namespace bb::pairing {
struct miller_lines;
} // namespace bb::pairing

namespace bb::srs::factories {

/**
 * A prover crs representation.
 */
template <typename Curve> class ProverCrs {
  public:
    virtual ~ProverCrs() = default;
    ;

    /**
     *  @brief Returns the monomial points in a form to be consumed by scalar_multiplication pippenger algorithm.
     */
    virtual typename Curve::AffineElement* get_monomial_points() = 0;
    virtual size_t get_monomial_size() const = 0;
};

template <typename Curve> class VerifierCrs {
  public:
    virtual ~VerifierCrs() = default;
};
template <> class VerifierCrs<curve::BN254> {
    using Curve = curve::BN254;

  public:
    virtual Curve::G2AffineElement get_g2x() const = 0;
    /**
     * @brief As the G_2 element of the CRS is fixed, we can precompute the operations performed on it during the
     * pairing algorithm to optimize pairing computations.
     */
    virtual bb::pairing::miller_lines const* get_precomputed_g2_lines() const = 0;
    /**
     *  @brief Returns the first G_1 element from the CRS, used by the Shplonk verifier to compute the final
     * commtiment.
     */
    virtual Curve::AffineElement get_first_g1() const = 0;
};

template <> class VerifierCrs<curve::Grumpkin> {
    using Curve = curve::Grumpkin;

  public:
    /**
     * @brief Returns the G_1 elements in the CRS after the pippenger point table has been applied on them
     *
     */
    virtual Curve::AffineElement* get_monomial_points() const = 0;
    virtual size_t get_monomial_size() const = 0;
    /**
     * @brief Returns the first G_1 element from the CRS, used by the Shplonk verifier to compute the final
     * commtiment.
     */
    virtual Curve::AffineElement get_first_g1() const = 0;
};

/**
 * A factory class to return the prover crs and verifier crs on request.
 * You can construct an empty placeholder factory, because composers need to be given a factory at construction time.
 */
template <typename Curve> class CrsFactory {
  public:
    CrsFactory() = default;
    CrsFactory(CrsFactory&& other) = default;
    virtual ~CrsFactory() = default;
    virtual std::shared_ptr<bb::srs::factories::ProverCrs<Curve>> get_prover_crs(size_t) { return nullptr; }
    virtual std::shared_ptr<bb::srs::factories::VerifierCrs<Curve>> get_verifier_crs([[maybe_unused]] size_t degree = 0)
    {
        return nullptr;
    }
};

} // namespace bb::srs::factories