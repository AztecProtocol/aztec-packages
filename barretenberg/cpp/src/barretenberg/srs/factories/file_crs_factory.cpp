#include "file_crs_factory.hpp"
#include "../io.hpp"
#include "barretenberg/ecc/curves/bn254/bn254.hpp"
#include "barretenberg/ecc/curves/bn254/g1.hpp"
#include "barretenberg/ecc/curves/bn254/pairing.hpp"
#include "barretenberg/ecc/curves/grumpkin/grumpkin.hpp"
#include "barretenberg/ecc/scalar_multiplication/point_table.hpp"
#include "barretenberg/ecc/scalar_multiplication/scalar_multiplication.hpp"

namespace bb::srs::factories {

FileVerifierCrs<curve::BN254>::FileVerifierCrs(std::string const& path, const size_t)
    : precomputed_g2_lines((bb::pairing::miller_lines*)(aligned_alloc(64, sizeof(bb::pairing::miller_lines) * 2)))
{
    using Curve = curve::BN254;
    auto point_buf = scalar_multiplication::point_table_alloc<Curve::AffineElement>(1);
    srs::IO<Curve>::read_transcript_g1(point_buf.get(), 1, path);
    srs::IO<curve::BN254>::read_transcript_g2(g2_x, path);
    bb::pairing::precompute_miller_lines(bb::g2::one, precomputed_g2_lines[0]);
    bb::pairing::precompute_miller_lines(g2_x, precomputed_g2_lines[1]);
    first_g1 = point_buf[0];
}

FileVerifierCrs<curve::BN254>::~FileVerifierCrs()
{
    aligned_free(precomputed_g2_lines);
}

FileVerifierCrs<curve::Grumpkin>::FileVerifierCrs(std::string const& path, const size_t num_points)
    : num_points(num_points)
{
    using Curve = curve::Grumpkin;
    monomials_ = scalar_multiplication::point_table_alloc<Curve::AffineElement>(num_points);
    srs::IO<Curve>::read_transcript_g1(monomials_.get(), num_points, path);
    scalar_multiplication::generate_pippenger_point_table<Curve>(monomials_.get(), monomials_.get(), num_points);
    first_g1 = monomials_[0];
};

curve::Grumpkin::AffineElement* FileVerifierCrs<curve::Grumpkin>::get_monomial_points() const
{
    return monomials_.get();
}

size_t FileVerifierCrs<curve::Grumpkin>::get_monomial_size() const
{
    return num_points;
}

template <typename Curve>
FileCrsFactory<Curve>::FileCrsFactory(std::string path, size_t initial_degree)
    : path_(std::move(path))
    , degree_(initial_degree)
{}

template <typename Curve>
std::shared_ptr<bb::srs::factories::ProverCrs<Curve>> FileCrsFactory<Curve>::get_prover_crs(size_t degree)
{
    if (degree != degree_ || !prover_crs_) {
        prover_crs_ = std::make_shared<FileProverCrs<Curve>>(degree, path_);
        degree_ = degree;
    }
    return prover_crs_;
}

template <typename Curve>
std::shared_ptr<bb::srs::factories::VerifierCrs<Curve>> FileCrsFactory<Curve>::get_verifier_crs(size_t degree)
{
    if (degree != degree_ || !verifier_crs_) {
        verifier_crs_ = std::make_shared<FileVerifierCrs<Curve>>(path_, degree);
        degree_ = degree;
    }
    return verifier_crs_;
}

template class FileProverCrs<curve::BN254>;
template class FileProverCrs<curve::Grumpkin>;
template class FileCrsFactory<curve::BN254>;
template class FileCrsFactory<curve::Grumpkin>;

} // namespace bb::srs::factories
