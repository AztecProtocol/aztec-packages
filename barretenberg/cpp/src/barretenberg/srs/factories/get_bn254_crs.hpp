#pragma once
#include <barretenberg/ecc/curves/bn254/g1.hpp>
#include <barretenberg/ecc/curves/bn254/g2.hpp>
#include <filesystem>
#include <fstream>
#include <ios>

namespace bb {
std::vector<g1::affine_element> get_bn254_g1_data(const std::filesystem::path& path,
                                                  size_t num_points,
                                                  bool allow_download = true);
g2::affine_element get_bn254_g2_data(const std::filesystem::path& path, bool allow_download = true);
} // namespace bb
