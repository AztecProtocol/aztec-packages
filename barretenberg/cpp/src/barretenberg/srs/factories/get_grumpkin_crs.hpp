#pragma once
#include <barretenberg/ecc/curves/bn254/g1.hpp>
#include <barretenberg/ecc/curves/grumpkin/grumpkin.hpp>
#include <filesystem>
#include <fstream>
#include <ios>

namespace bb {

std::vector<curve::Grumpkin::AffineElement> get_grumpkin_g1_data(const std::filesystem::path& path,
                                                                 size_t num_points,
                                                                 bool allow_download = true);

}
