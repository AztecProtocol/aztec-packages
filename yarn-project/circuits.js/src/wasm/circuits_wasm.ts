import { numToUInt32BE } from '@aztec/foundation/serialize';
import { fileURLToPath } from '@aztec/foundation/url';
import { IWasmModule, WasmModule } from '@aztec/foundation/wasm';

import isNode from 'detect-node';
import { readFile } from 'fs/promises';
import { dirname, join } from 'path';

import { Crs } from '../crs/index.js';

const NAME = '/aztec3-circuits';
const CODE_PATH = isNode
  ? join(dirname(fileURLToPath(import.meta.url)), `../../resources/${NAME}.wasm`)
  : `${NAME}.wasm`;
const MAX_SRS_POINTS = 50000;

/**
 * Get the text of a binary file, either locally or on the web.
 * @param path - Path to the WASM binary.
 * @returns The binary buffer.
 */
async function fetchBinary(path: string) {
  if (isNode) {
    return await readFile(path);
  } else {
    const res = await fetch(path);
    return Buffer.from(await res.arrayBuffer());
  }
}
const BB_JS_NYI_ERROR =
  'NOT YET IMPLEMENTED - needed for proofs, plan is to use barretenberg.js from NPM for proofs. See https://github.com/AztecProtocol/aztec-packages/issues/781';
/**
 * A low-level wrapper for an instance of Aztec3 Circuits WASM.
 */
export class CircuitsWasm implements IWasmModule {
  static instance: Promise<CircuitsWasm>;

  /**
   * Creates a Circuits module from a properly configured WasmModule.
   * Not meant for public use.
   *
   * @param wasm - The module configured in CircuitsWasm.new().
   */
  private constructor(private wasm: WasmModule) {}

  /**
   * Get a singleton instance of the module.
   * @returns The singleton.
   */
  public static get(): Promise<CircuitsWasm> {
    if (!this.instance) this.instance = CircuitsWasm.new();
    return this.instance;
  }

  /**
   * Create and initialize a Circuits module. Not meant for public use.
   *
   * Has 30 pages by default. 30*2**16 \> 1mb stack size plus other overheads.
   * 8192 maximum by default. 512mb.
   * @param initial - Initial memory pages.
   * @param maximum - Max memory pages.
   * @param loggerName - The logger name.
   * @returns The wrapper.
   */
  private static async new(initial = 30, maximum = 8192, loggerName = 'aztec:wasm'): Promise<CircuitsWasm> {
    const wasm = new WasmModule(
      await fetchBinary(CODE_PATH),
      module => ({
        /**
         * Log a string from wasm.
         * @param addr - The string address to log.
         */
        logstr(addr: number) {
          const rawStr = wasm.getMemoryAsString(addr);
          const m = wasm.getMemory();
          const str = `${rawStr} (mem: ${(m.length / (1024 * 1024)).toFixed(2)}MB)`;
          if (str.startsWith('abort: ') || str.startsWith('important: ')) {
            // we explicitly want to route to console.log for every abort message to not miss them:
            // eslint-disable-next-line no-console
            console.log(str);
          }
          wasm.getLogger()(str);
        },
        memory: module.getRawMemory(),
        // eslint-disable-next-line camelcase
        set_data: () => {
          wasm.getLogger()('set_data: NYI');
        },
        // eslint-disable-next-line camelcase
        get_data: () => {
          throw new Error(BB_JS_NYI_ERROR);
        },
      }),
      loggerName,
    );
    await wasm.init(initial, maximum);
    await CircuitsWasm.initializeSrs(wasm);
    return new CircuitsWasm(wasm);
  }

  /**
   * Ensures the global SRS is initialized.
   * Currently used in VK serialization and will be used in proofs.
   * TODO(AD): proof should use external bb.js
   * TODO(AD): revisit when SRS should be initialized
   * @param wasm - The WASM module.
   */
  private static async initializeSrs(wasm: WasmModule) {
    const crs = new Crs(MAX_SRS_POINTS);
    await crs.init();
    const g1Buf = wasm.call('bbmalloc', crs.getG1Data().length);
    wasm.writeMemory(g1Buf, crs.getG1Data());
    const g1SizeBuf = wasm.call('bbmalloc', 4);
    wasm.writeMemory(g1SizeBuf, numToUInt32BE(crs.numPoints));
    const g2Buf = wasm.call('bbmalloc', crs.getG2Data().length);
    wasm.writeMemory(g2Buf, crs.getG2Data());
    wasm.call('srs_init_srs', g1Buf, g1SizeBuf, g2Buf);
    wasm.call('bbfree', g1Buf);
    wasm.call('bbfree', g1SizeBuf);
    wasm.call('bbfree', g2Buf);
  }

  /**
   * Retrieve the exports object of the CircuitsWasm module.
   *
   * @returns An object containing exported functions and properties.
   */
  public exports(): any {
    return this.wasm.exports();
  }

  /**
   * Get a slice of memory between two addresses.
   * @param start - The start address.
   * @param end - The end address.
   * @returns A Uint8Array view of memory.
   */
  public getMemorySlice(start: number, end: number) {
    return this.wasm.getMemorySlice(start, end);
  }

  /**
   * Write data into the heap.
   * @param arr - The data to write.
   * @param offset - The address to write data at.
   */
  public writeMemory(offset: number, arr: Uint8Array) {
    this.wasm.writeMemory(offset, arr);
  }

  /**
   * Get memory as string.
   * @param offset - The address to get null-terminated string data from.
   * @returns JS string.
   */
  public getMemoryAsString(offset: number) {
    return this.wasm.getMemoryAsString(offset);
  }

  /**
   * Calls into the WebAssembly.
   * @param name - The method name.
   * @param args - The arguments to the method.
   * @returns The numeric integer or address result.
   */
  public call(name: string, ...args: any): number {
    return this.wasm.call(name, ...args);
  }
}
