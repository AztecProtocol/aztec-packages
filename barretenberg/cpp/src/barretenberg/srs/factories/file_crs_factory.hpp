#pragma once
#include "../io.hpp"
#include "barretenberg/ecc/curves/bn254/bn254.hpp"
#include "barretenberg/ecc/curves/grumpkin/grumpkin.hpp"
#include "barretenberg/ecc/scalar_multiplication/point_table.hpp"
#include "barretenberg/ecc/scalar_multiplication/scalar_multiplication.hpp"
#include "crs_factory.hpp"
#include <cstddef>
#include <utility>

namespace bb::srs::factories {

/**
 * Create reference strings given a path to a directory of transcript files.
 */
template <typename Curve> class FileCrsFactory : public CrsFactory<Curve> {
  public:
    FileCrsFactory(std::string path, size_t initial_degree = 0);
    FileCrsFactory(FileCrsFactory&& other) = default;

    std::shared_ptr<bb::srs::factories::ProverCrs<Curve>> get_prover_crs(size_t degree) override;

    std::shared_ptr<bb::srs::factories::VerifierCrs<Curve>> get_verifier_crs(size_t degree = 0) override;

  private:
    std::string path_;
    size_t degree_;
    std::shared_ptr<bb::srs::factories::ProverCrs<Curve>> prover_crs_;
    std::shared_ptr<bb::srs::factories::VerifierCrs<Curve>> verifier_crs_;
};

template <typename Curve> class FileProverCrs : public ProverCrs<Curve> {
  public:
    FileProverCrs(const size_t num_points, std::string const& path)
        : num_points(num_points)
    {
        monomials_ = scalar_multiplication::point_table_alloc<typename Curve::AffineElement>(num_points);

        srs::IO<Curve>::read_transcript_g1(monomials_.get(), num_points, path);
        scalar_multiplication::generate_pippenger_point_table<Curve>(monomials_.get(), monomials_.get(), num_points);
    };

    typename Curve::AffineElement* get_monomial_points() { return monomials_.get(); }

    [[nodiscard]] size_t get_monomial_size() const { return num_points; }

  private:
    size_t num_points;
    std::shared_ptr<typename Curve::AffineElement[]> monomials_;
};

template <typename Curve> class FileVerifierCrs : public VerifierCrs<Curve> {
  public:
    FileVerifierCrs(std::string const& path, const size_t num_points);
};

template <> class FileVerifierCrs<curve::BN254> : public VerifierCrs<curve::BN254> {
    using Curve = curve::BN254;

  public:
    FileVerifierCrs(std::string const& path, const size_t num_points = 0);
    virtual ~FileVerifierCrs();
    Curve::G2AffineElement get_g2x() const override { return g2_x; };
    pairing::miller_lines const* get_precomputed_g2_lines() const override { return precomputed_g2_lines; };
    Curve::AffineElement get_first_g1() const override { return first_g1; };

  private:
    Curve::AffineElement first_g1;
    Curve::G2AffineElement g2_x;
    pairing::miller_lines* precomputed_g2_lines;
};

template <> class FileVerifierCrs<curve::Grumpkin> : public VerifierCrs<curve::Grumpkin> {
    using Curve = curve::Grumpkin;

  public:
    FileVerifierCrs(std::string const& path, const size_t num_points);
    virtual ~FileVerifierCrs() = default;
    Curve::AffineElement* get_monomial_points() const override;
    size_t get_monomial_size() const override;
    Curve::AffineElement get_first_g1() const override { return first_g1; };

  private:
    Curve::AffineElement first_g1;
    size_t num_points;
    std::shared_ptr<Curve::AffineElement[]> monomials_;
};

} // namespace bb::srs::factories
