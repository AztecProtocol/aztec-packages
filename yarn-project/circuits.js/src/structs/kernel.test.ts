import { expectSerializeToMatchSnapshot } from '../tests/expectSerialize.js';
import { makePreviousKernelData, makePrivateKernelInputs } from '../tests/factories.js';
import { CircuitsWasm } from '../wasm/circuits_wasm.js';

describe('structs/kernel', () => {
  it(`serializes and prints previous_kernel_data`, async () => {
    const wasm = await CircuitsWasm.new();
    const previousKernelData = makePreviousKernelData();
    await expectSerializeToMatchSnapshot(
      previousKernelData.toBuffer(),
      'abis__test_roundtrip_serialize_previous_kernel_data',
      wasm,
    );
  });

  it(`serializes and prints private_kernel_inputs`, async () => {
    const wasm = await CircuitsWasm.new();
    const kernelInputs = makePrivateKernelInputs();
    await expectSerializeToMatchSnapshot(
      kernelInputs.toBuffer(),
      'abis__test_roundtrip_serialize_private_kernel_inputs',
      wasm,
    );
  });
});
