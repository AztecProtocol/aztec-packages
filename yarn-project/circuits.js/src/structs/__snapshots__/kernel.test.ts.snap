// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`structs/kernel serializes and prints previous_kernel_data 1`] = `
"public_inputs: end:
aggregation_object:
P0: { 0x1, 0x2 }
P1: { 0x101, 0x102 }
public_inputs: [
  0: 0x3
  1: 0x4
  2: 0x5
  3: 0x6
]
proof_witness_indices: [
  0: 7
  1: 8
  2: 9
  3: 10
  4: 11
  5: 12
has_data: 0

private_call_count: 0xd
new_commitments:
[ 0x101 0x102 0x103 0x104 ]
new_nullifiers:
[ 0x201 0x202 0x203 0x204 ]
private_call_stack:
[ 0x301 0x302 0x303 0x304 0x305 0x306 0x307 0x308 ]
public_call_stack:
[ 0x401 0x402 0x403 0x404 0x405 0x406 0x407 0x408 ]
l1_msg_stack:
[ 0x501 0x502 0x503 0x504 ]
new_contracts:
[ contract_address: 0x601
portal_contract_address: 0x202020202020202020202020202020202020202
function_tree_root: 0x603
 ]
optionally_revealed_data:
[ call_stack_item_hash: 0x701
function_data:
function_selector: 702
is_private: 1
is_constructor: 1

emitted_events:
[ 0x801 0x802 0x803 0x804 ]
vk_hash: 0x703
portal_contract_address: 0x404040404040404040404040404040404040404
pay_fee_from_l1: 1
pay_fee_from_public_l2: 0
called_from_l1: 1
called_from_public_l2: 0
 call_stack_item_hash: 0x702
function_data:
function_selector: 703
is_private: 1
is_constructor: 1

emitted_events:
[ 0x802 0x803 0x804 0x805 ]
vk_hash: 0x704
portal_contract_address: 0x505050505050505050505050505050505050505
pay_fee_from_l1: 1
pay_fee_from_public_l2: 0
called_from_l1: 1
called_from_public_l2: 0
 call_stack_item_hash: 0x703
function_data:
function_selector: 704
is_private: 1
is_constructor: 1

emitted_events:
[ 0x803 0x804 0x805 0x806 ]
vk_hash: 0x705
portal_contract_address: 0x606060606060606060606060606060606060606
pay_fee_from_l1: 1
pay_fee_from_public_l2: 0
called_from_l1: 1
called_from_public_l2: 0
 call_stack_item_hash: 0x704
function_data:
function_selector: 705
is_private: 1
is_constructor: 1

emitted_events:
[ 0x804 0x805 0x806 0x807 ]
vk_hash: 0x706
portal_contract_address: 0x707070707070707070707070707070707070707
pay_fee_from_l1: 1
pay_fee_from_public_l2: 0
called_from_l1: 1
called_from_public_l2: 0
 ]

constants:
old_tree_roots: private_data_tree_root: 0x101
nullifier_tree_root: 0x102
contract_tree_root: 0x103
private_kernel_vk_tree_root: 0x104

tx_context: is_fee_payment_tx: 0
is_rebate_payment_tx: 0
is_contract_deployment_tx: 1
contract_deployment_data: 
constructor_vk_hash: 0x105
function_tree_root: 0x106
contract_address_salt: 0x107
portal_contract_address: 0x808080808080808080808080808080808080808


is_private: 1

proof: [ 81 81 81 81 81 81 81 81 81 81 81 81 81 81 81 81 ]
vk: 0x12ad40
vk_index: 66
vk_path: [ 0x1000 0x1001 0x1002 ]
"
`;

exports[`structs/kernel serializes and prints private_kernel_inputs 1`] = `
"signed_tx_request:
tx_request:
from: 0x1
to: 0x11
function_data: function_selector: 257
is_private: 1
is_constructor: 1

args: [ 0x201 0x202 0x203 0x204 0x205 0x206 0x207 0x208 ]
nonce: 0x301
tx_context: is_fee_payment_tx: 0
is_rebate_payment_tx: 0
is_contract_deployment_tx: 1
contract_deployment_data: 
constructor_vk_hash: 0x401
function_tree_root: 0x402
contract_address_salt: 0x403
portal_contract_address: 0x404040404040404040404040404040404040404

chain_id: 0x501

signature: { 0101010101010101010101010101010101010101010101010101010101010101, 0202020202020202020202020202020202020202020202020202020202020202 }

previous_kernel:
public_inputs: end:
aggregation_object:
P0: { 0x1001, 0x1002 }
P1: { 0x1101, 0x1102 }
public_inputs: [
  0: 0x1003
  1: 0x1004
  2: 0x1005
  3: 0x1006
]
proof_witness_indices: [
  0: 4103
  1: 4104
  2: 4105
  3: 4106
  4: 4107
  5: 4108
has_data: 0

private_call_count: 0x100d
new_commitments:
[ 0x1101 0x1102 0x1103 0x1104 ]
new_nullifiers:
[ 0x1201 0x1202 0x1203 0x1204 ]
private_call_stack:
[ 0x1301 0x1302 0x1303 0x1304 0x1305 0x1306 0x1307 0x1308 ]
public_call_stack:
[ 0x1401 0x1402 0x1403 0x1404 0x1405 0x1406 0x1407 0x1408 ]
l1_msg_stack:
[ 0x1501 0x1502 0x1503 0x1504 ]
new_contracts:
[ contract_address: 0x1601
portal_contract_address: 0x202020202020202020202020202020202020202
function_tree_root: 0x1603
 ]
optionally_revealed_data:
[ call_stack_item_hash: 0x1701
function_data:
function_selector: 1702
is_private: 1
is_constructor: 1

emitted_events:
[ 0x1801 0x1802 0x1803 0x1804 ]
vk_hash: 0x1703
portal_contract_address: 0x404040404040404040404040404040404040404
pay_fee_from_l1: 1
pay_fee_from_public_l2: 0
called_from_l1: 1
called_from_public_l2: 0
 call_stack_item_hash: 0x1702
function_data:
function_selector: 1703
is_private: 1
is_constructor: 1

emitted_events:
[ 0x1802 0x1803 0x1804 0x1805 ]
vk_hash: 0x1704
portal_contract_address: 0x505050505050505050505050505050505050505
pay_fee_from_l1: 1
pay_fee_from_public_l2: 0
called_from_l1: 1
called_from_public_l2: 0
 call_stack_item_hash: 0x1703
function_data:
function_selector: 1704
is_private: 1
is_constructor: 1

emitted_events:
[ 0x1803 0x1804 0x1805 0x1806 ]
vk_hash: 0x1705
portal_contract_address: 0x606060606060606060606060606060606060606
pay_fee_from_l1: 1
pay_fee_from_public_l2: 0
called_from_l1: 1
called_from_public_l2: 0
 call_stack_item_hash: 0x1704
function_data:
function_selector: 1705
is_private: 1
is_constructor: 1

emitted_events:
[ 0x1804 0x1805 0x1806 0x1807 ]
vk_hash: 0x1706
portal_contract_address: 0x707070707070707070707070707070707070707
pay_fee_from_l1: 1
pay_fee_from_public_l2: 0
called_from_l1: 1
called_from_public_l2: 0
 ]

constants:
old_tree_roots: private_data_tree_root: 0x1101
nullifier_tree_root: 0x1102
contract_tree_root: 0x1103
private_kernel_vk_tree_root: 0x1104

tx_context: is_fee_payment_tx: 0
is_rebate_payment_tx: 0
is_contract_deployment_tx: 1
contract_deployment_data: 
constructor_vk_hash: 0x1105
function_tree_root: 0x1106
contract_address_salt: 0x1107
portal_contract_address: 0x808080808080808080808080808080808080808


is_private: 1

proof: [ 81 81 81 81 81 81 81 81 81 81 81 81 81 81 81 81 ]
vk: 0x12cd80
vk_index: 66
vk_path: [ 0x1000 0x1001 0x1002 ]

private_call:
call_stack_item:
contract_address: 0x2001
function_data: function_selector: 8194
is_private: 1
is_constructor: 1

public_inputs: call_context: msg_sender: 0x2012
storage_contract_address: 0x2013
tx_origin: 0x2014
is_delegate_call: 1
is_static_call: 1
is_contract_deployment: 1

args: [ 0x2111 0x2112 0x2113 0x2114 0x2115 0x2116 0x2117 0x2118 ]
return_values: [ 0x2311 0x2312 0x2313 0x2314 ]
emitted_events: [ 0x2211 0x2212 0x2213 0x2214 ]
new_commitments: [ 0x2411 0x2412 0x2413 0x2414 ]
new_nullifiers: [ 0x2511 0x2512 0x2513 0x2514 ]
private_call_stack: [ 0x2611 0x2612 0x2613 0x2614 ]
public_call_stack: [ 0x2711 0x2712 0x2713 0x2714 ]
l1_msg_stack: [ 0x2811 0x2812 ]
historic_private_data_tree_root: 0x3011
historic_nullifier_tree_root: 0x3111
contract_deployment_data: constructor_vk_hash: 0x1
function_tree_root: 0x2
contract_address_salt: 0x3
portal_contract_address: 0x4

contract_deployment_data: constructor_vk_hash: 0x1
function_tree_root: 0x2
contract_address_salt: 0x3
portal_contract_address: 0x4



private_call_stack_preimages:
[ contract_address: 0x2011
function_data: function_selector: 2012
is_private: 1
is_constructor: 1

public_inputs: call_context: msg_sender: 0x2022
storage_contract_address: 0x2023
tx_origin: 0x2024
is_delegate_call: 1
is_static_call: 1
is_contract_deployment: 1

args: [ 0x2121 0x2122 0x2123 0x2124 0x2125 0x2126 0x2127 0x2128 ]
return_values: [ 0x2321 0x2322 0x2323 0x2324 ]
emitted_events: [ 0x2221 0x2222 0x2223 0x2224 ]
new_commitments: [ 0x2421 0x2422 0x2423 0x2424 ]
new_nullifiers: [ 0x2521 0x2522 0x2523 0x2524 ]
private_call_stack: [ 0x2621 0x2622 0x2623 0x2624 ]
public_call_stack: [ 0x2721 0x2722 0x2723 0x2724 ]
l1_msg_stack: [ 0x2821 0x2822 ]
historic_private_data_tree_root: 0x3021
historic_nullifier_tree_root: 0x3121
contract_deployment_data: constructor_vk_hash: 0x1
function_tree_root: 0x2
contract_address_salt: 0x3
portal_contract_address: 0x4

contract_deployment_data: constructor_vk_hash: 0x1
function_tree_root: 0x2
contract_address_salt: 0x3
portal_contract_address: 0x4


 contract_address: 0x2012
function_data: function_selector: 2013
is_private: 1
is_constructor: 1

public_inputs: call_context: msg_sender: 0x2023
storage_contract_address: 0x2024
tx_origin: 0x2025
is_delegate_call: 1
is_static_call: 1
is_contract_deployment: 1

args: [ 0x2122 0x2123 0x2124 0x2125 0x2126 0x2127 0x2128 0x2129 ]
return_values: [ 0x2322 0x2323 0x2324 0x2325 ]
emitted_events: [ 0x2222 0x2223 0x2224 0x2225 ]
new_commitments: [ 0x2422 0x2423 0x2424 0x2425 ]
new_nullifiers: [ 0x2522 0x2523 0x2524 0x2525 ]
private_call_stack: [ 0x2622 0x2623 0x2624 0x2625 ]
public_call_stack: [ 0x2722 0x2723 0x2724 0x2725 ]
l1_msg_stack: [ 0x2822 0x2823 ]
historic_private_data_tree_root: 0x3022
historic_nullifier_tree_root: 0x3122
contract_deployment_data: constructor_vk_hash: 0x1
function_tree_root: 0x2
contract_address_salt: 0x3
portal_contract_address: 0x4

contract_deployment_data: constructor_vk_hash: 0x1
function_tree_root: 0x2
contract_address_salt: 0x3
portal_contract_address: 0x4


 contract_address: 0x2013
function_data: function_selector: 2014
is_private: 1
is_constructor: 1

public_inputs: call_context: msg_sender: 0x2024
storage_contract_address: 0x2025
tx_origin: 0x2026
is_delegate_call: 1
is_static_call: 1
is_contract_deployment: 1

args: [ 0x2123 0x2124 0x2125 0x2126 0x2127 0x2128 0x2129 0x212a ]
return_values: [ 0x2323 0x2324 0x2325 0x2326 ]
emitted_events: [ 0x2223 0x2224 0x2225 0x2226 ]
new_commitments: [ 0x2423 0x2424 0x2425 0x2426 ]
new_nullifiers: [ 0x2523 0x2524 0x2525 0x2526 ]
private_call_stack: [ 0x2623 0x2624 0x2625 0x2626 ]
public_call_stack: [ 0x2723 0x2724 0x2725 0x2726 ]
l1_msg_stack: [ 0x2823 0x2824 ]
historic_private_data_tree_root: 0x3023
historic_nullifier_tree_root: 0x3123
contract_deployment_data: constructor_vk_hash: 0x1
function_tree_root: 0x2
contract_address_salt: 0x3
portal_contract_address: 0x4

contract_deployment_data: constructor_vk_hash: 0x1
function_tree_root: 0x2
contract_address_salt: 0x3
portal_contract_address: 0x4


 contract_address: 0x2014
function_data: function_selector: 2015
is_private: 1
is_constructor: 1

public_inputs: call_context: msg_sender: 0x2025
storage_contract_address: 0x2026
tx_origin: 0x2027
is_delegate_call: 1
is_static_call: 1
is_contract_deployment: 1

args: [ 0x2124 0x2125 0x2126 0x2127 0x2128 0x2129 0x212a 0x212b ]
return_values: [ 0x2324 0x2325 0x2326 0x2327 ]
emitted_events: [ 0x2224 0x2225 0x2226 0x2227 ]
new_commitments: [ 0x2424 0x2425 0x2426 0x2427 ]
new_nullifiers: [ 0x2524 0x2525 0x2526 0x2527 ]
private_call_stack: [ 0x2624 0x2625 0x2626 0x2627 ]
public_call_stack: [ 0x2724 0x2725 0x2726 0x2727 ]
l1_msg_stack: [ 0x2824 0x2825 ]
historic_private_data_tree_root: 0x3024
historic_nullifier_tree_root: 0x3124
contract_deployment_data: constructor_vk_hash: 0x1
function_tree_root: 0x2
contract_address_salt: 0x3
portal_contract_address: 0x4

contract_deployment_data: constructor_vk_hash: 0x1
function_tree_root: 0x2
contract_address_salt: 0x3
portal_contract_address: 0x4


 ]
proof:
[ 51 51 51 51 51 51 51 51 51 51 51 51 51 51 51 51 ]
vk:
0x1357c0
function_leaf_membership_witness:
leaf_index: 8241
sibling_path: [ 0x2031 0x2032 0x2033 0x2034 ]

contract_leaf_membership_witness:
leaf_index: 8225
sibling_path: [ 0x2021 0x2022 0x2023 0x2024 ]

portal_contract_address: 0x4141414141414141414141414141414141414141

"
`;
